### PR TITLE
Support large files on Windows platforms

### PR DESCRIPTION
**Pull request by TJOSERAFAEL**
### Large files support

As described on this issue https://github.com/wazuh/wazuh/issues/3178, the `w_ftell` function causes an error when reading files > 4GB due to the binary being compiled for 32 bit.

As on the macro definition for `w_ftell` https://github.com/wazuh/wazuh/blob/3.10/src/headers/shared.h#L190 an `merror_exit` causes the agent to stop

### Remediation 

- Change `merror_exit` call inside `w_ftell` function to `merror`.
- Return `-1` if `ftell` fails.
- Use `_ftelli64` for Windows platforms (https://docs.microsoft.com/es-es/cpp/c-runtime-library/reference/ftell-ftelli64?view=vs-2019).
- On the logcollector reader code, use 64 bit variables.

### Testing

Tested 8 GB file successfully on:

- Windows XP SP3 (32 bits)
- Windows 2003 Server (32 bits)
- Windows 7
- Windows 2012 Server

- Remaining:
  - Launch compilation for HP-UX, AIX systems

### CODE REVIEW

File: src/logcollector/read_json.c [hunk 1]
```diff
@@ -36,13 +36,23 @@ void *read_json(logreader *lf, int *rc, int drop_it) {
         rbytes = w_ftell(lf->fp) - offset;
         lines++;
 
+        /* Flow control */
+        if ( rbytes <= 0) {
```
Comment (vikman90) [inline src/logcollector/read_json.c]: Please remove the space between `(` and `rbytes`.

File: src/logcollector/read_json.c [hunk 2]
```diff
@@ -106,15 +116,29 @@ void *read_json(logreader *lf, int *rc, int drop_it) {
...
                 mdebug2("Large message size from file '%s' (length = %ld): '%.*s'...", lf->file, rbytes, sample_log_length, str);
+                #endif
+                
             }
 
             for (offset += rbytes; fgets(str, OS_MAXSTR - 2, lf->fp) != NULL; offset += rbytes) {
                 rbytes = w_ftell(lf->fp) - offset;
 
+                /* Flow control */
+                if ( rbytes <= 0) {
```
Comment (vikman90) [inline src/logcollector/read_json.c]: Same whitespace here.

File: src/logcollector/read_multiline.c [hunk 3]
```diff
@@ -97,15 +107,28 @@ void *read_multiline(logreader *lf, int *rc, int drop_it) {
...
+                #else
                 mdebug2("Large message size from file '%s' (length = %ld): '%.*s'...", lf->file, rbytes, sample_log_length, str);
+                #endif
             }
 
             for (offset += rbytes; fgets(str, OS_MAXSTR - 2, lf->fp) != NULL; offset += rbytes) {
                 rbytes = w_ftell(lf->fp) - offset;
 
+                /* Flow control */
+                if ( rbytes <= 0) {
```
Comment (vikman90) [inline src/logcollector/read_multiline.c]: Another whitespace here.

File: src/logcollector/read_syslog.c [hunk 4]
```diff
@@ -34,13 +39,23 @@ void *read_syslog(logreader *lf, int *rc, int drop_it) {
         rbytes = w_ftell(lf->fp) - offset;
         lines++;
 
+        /* Flow control */
+        if ( rbytes <= 0) {
```
Comment (vikman90) [inline src/logcollector/read_syslog.c]: Please fix this whitespace and the rest of similar lines.

File: src/logcollector/read_audit.c [hunk 5]
```diff
@@ -48,22 +48,34 @@ void *read_audit(logreader *lf, int *rc, int drop_it) {
...
         lines++;
 
         if (buffer[rbytes - 1] == '\n') {
             buffer[rbytes - 1] = '\0';
 
-            if ((long)strlen(buffer) != rbytes - 1)
+            if ((int64_t)strlen(buffer) != rbytes - 1)
             {
+                #ifdef WIN32
+                mdebug2("Line in '%s' contains some zero-bytes (valid=%lld / total=%lld). Dropping line.", lf->file, (int64_t)strlen(buffer), rbytes - 1);
```
Comment (vikman90) [inline src/logcollector/read_audit.c]: Can we simplify this bunch of repeated messages?
For instance:
```c
#ifdef WIN32
#define FTELL_TT "lld"
#else
#define FTELL_TT "ld"
#endif
// (...)
mdebug2("Line in '%s' contains some zero-bytes (valid=%" FTELL_TT " / total=%" FTELL_TT "). Dropping line.", lf->file, strlen(buffer), rbytes - 1);
```

Same fix if `strlen(buffer)` produces a warning.

**Comment by chemamartinez** [comment]
Warning compilations on MAC OS X agent:

```
shared/file_op.c:915:125: warning: format specifies type 'long' but the argument has type 'long long' [-Wformat]
                mdebug2("Line contains some zero-bytes (valid=" FTELL_TT "/ total=" FTELL_TT ").", FTELL_INT64 strlen(str), rbytes - 1);
                                                                                                                            ^~~~~~~~~~
./headers/debug_op.h:42:73: note: expanded from macro 'mdebug2'
#define mdebug2(msg, ...) _mdebug2(__FILE__, __LINE__, __func__, msg, ##__VA_ARGS__)
                                                                        ^~~~~~~~~~~
    CC shared/help.o
    CC shared/json-queue.o
    CC shared/json_op.o
    CC shared/labels_op.o
    CC shared/list_op.o
    CC shared/math_op.o
    CC shared/mem_op.o
    CC shared/mq_op.o
1 warning generated.
    CC shared/notify_op.o
    CC shared/os_utils.o
    CC shared/privsep_op.o
    CC shared/pthreads_op.o
    CC shared/queue_op.o
    CC shared/randombytes.o
    CC shared/read-agents.o
    CC shared/read-alert.o
    CC shared/regex_op.o
    CC shared/report_op.o
    CC shared/request_op.o
    CC shared/rootcheck_op.o
    CC shared/rules_op.o
    CC shared/sig_op.o
    CC shared/store_op.o
    CC shared/string_op.o
    CC shared/syscheck_op.o
    CC shared/time_op.o
    CC shared/url.o
    CC shared/validate_op.o
    CC shared/vector_op.o
    CC shared/version_op.o
    CC shared/wait_op.o
    CC shared/wazuhdb_op.o
    CC shared/yaml2json.o
    CC os_net/os_net.o
    CC os_regex/os_match.o
    CC os_regex/os_match_compile.o
    CC os_regex/os_match_execute.o
    CC os_regex/os_match_free_pattern.o
    CC os_regex/os_regex.o
    CC os_regex/os_regex_compile.o
    CC os_regex/os_regex_execute.o
    CC os_regex/os_regex_free_pattern.o
    CC os_regex/os_regex_maps.o
    CC os_regex/os_regex_match.o
    CC os_regex/os_regex_startswith.o
    CC os_regex/os_regex_str.o
    CC os_regex/os_regex_strbreak.o
    CC os_xml/os_xml.o
    CC os_xml/os_xml_access.o
    CC os_xml/os_xml_node_access.o
    CC os_xml/os_xml_variables.o
    CC os_xml/os_xml_writer.o
    CC os_zlib/os_zlib.o
    CC libwazuhext.dylib
ld: warning: option -noall_load is obsolete and being ignored
    CC addagent/validate.o
    CC os_auth/main-client.o
    CC os_auth/ssl.o
    CC os_auth/check_cert.o
    CC logcollector/config.o
    CC logcollector/lccom.o
    CC logcollector/logcollector.o
    CC logcollector/main.o
    CC logcollector/read_audit.o
    CC logcollector/read_command.o
    CC logcollector/read_djb_multilog.o
logcollector/read_audit.c:73:162: warning: format specifies type 'long' but the argument has type 'long long' [-Wformat]
  ...some zero-bytes (valid=" FTELL_TT " / total=" FTELL_TT "). Dropping line.", lf->file, FTELL_INT64 strlen(buffer), rbytes - 1);
                                                                                                                       ^~~~~~~~~~
./headers/debug_op.h:42:73: note: expanded from macro 'mdebug2'
#define mdebug2(msg, ...) _mdebug2(__FILE__, __LINE__, __func__, msg, ##__VA_ARGS__)
                                                                        ^~~~~~~~~~~
1 warning generated.
    CC logcollector/read_fullcommand.o
    CC logcollector/read_json.o
    CC logcollector/read_mssql_log.o
logcollector/read_json.c:50:159: warning: format specifies type 'long' but the argument has type 'long long' [-Wformat]
  ...contains some zero-bytes (valid=" FTELL_TT " / total=" FTELL_TT "). Dropping line.", lf->file, FTELL_INT64 strlen(str), rbytes - 1);
                                                                                                                             ^~~~~~~~~~
./headers/debug_op.h:42:73: note: expanded from macro 'mdebug2'
#define mdebug2(msg, ...) _mdebug2(__FILE__, __LINE__, __func__, msg, ##__VA_ARGS__)
                                                                        ^~~~~~~~~~~
logcollector/read_json.c:114:106: warning: format specifies type 'long' but the argument has type 'int64_t' (aka 'long long') [-Wformat]
                merror("Large message size from file '%s' (length = " FTELL_TT "): '%.*s'...", lf->file, rbytes, sample_log_length, str);
                                                                                                         ^~~~~~
./headers/debug_op.h:44:71: note: expanded from macro 'merror'
#define merror(msg, ...) _merror(__FILE__, __LINE__, __func__, msg, ##__VA_ARGS__)
                                                                      ^~~~~~~~~~~
logcollector/read_json.c:117:107: warning: format specifies type 'long' but the argument has type 'int64_t' (aka 'long long') [-Wformat]
                mdebug2("Large message size from file '%s' (length = " FTELL_TT "): '%.*s'...", lf->file, rbytes, sample_log_length, str);
                                                                                                          ^~~~~~
./headers/debug_op.h:42:73: note: expanded from macro 'mdebug2'
#define mdebug2(msg, ...) _mdebug2(__FILE__, __LINE__, __func__, msg, ##__VA_ARGS__)
                                                                        ^~~~~~~~~~~
    CC logcollector/read_multiline.o
3 warnings generated.
    CC logcollector/read_mysql_log.o
    CC logcollector/read_nmapg.o
logcollector/read_multiline.c:52:159: warning: format specifies type 'long' but the argument has type 'long long' [-Wformat]
  ...contains some zero-bytes (valid=" FTELL_TT " / total=" FTELL_TT "). Dropping line.", lf->file, FTELL_INT64 strlen(str), rbytes - 1);
                                                                                                                             ^~~~~~~~~~
./headers/debug_op.h:42:73: note: expanded from macro 'mdebug2'
#define mdebug2(msg, ...) _mdebug2(__FILE__, __LINE__, __func__, msg, ##__VA_ARGS__)
                                                                        ^~~~~~~~~~~
logcollector/read_multiline.c:105:106: warning: format specifies type 'long' but the argument has type 'int64_t' (aka 'long long')
      [-Wformat]
                merror("Large message size from file '%s' (length = " FTELL_TT "): '%.*s'...", lf->file, rbytes, sample_log_length, str);
                                                                                                         ^~~~~~
./headers/debug_op.h:44:71: note: expanded from macro 'merror'
#define merror(msg, ...) _merror(__FILE__, __LINE__, __func__, msg, ##__VA_ARGS__)
                                                                      ^~~~~~~~~~~
logcollector/read_multiline.c:108:107: warning: format specifies type 'long' but the argument has type 'int64_t' (aka 'long long')
      [-Wformat]
                mdebug2("Large message size from file '%s' (length = " FTELL_TT "): '%.*s'...", lf->file, rbytes, sample_log_length, str);
                                                                                                          ^~~~~~
./headers/debug_op.h:42:73: note: expanded from macro 'mdebug2'
#define mdebug2(msg, ...) _mdebug2(__FILE__, __LINE__, __func__, msg, ##__VA_ARGS__)
                                                                        ^~~~~~~~~~~
3 warnings generated.
```

**Comment by TJOSERAFAEL** [comment]
Hi @chemamartinez ,

fixed MacOS compilation warnings on this commit: https://github.com/wazuh/wazuh/pull/3248/commits/cabfe6dfe9ef3958a0b1e59b327fe3e0799df700

**Comment by TJOSERAFAEL** [comment]
### Compilation after fix https://github.com/wazuh/wazuh/commit/cabfe6dfe9ef3958a0b1e59b327fe3e0799df700

- [x] Debian
- [x] Windows
- [x] MacOS
- [x] Solaris 10

Related warnings are gone.

**Comment by TJOSERAFAEL** [comment]
### Valgrind report

```
==12582== 
==12582== HEAP SUMMARY:
==12582==     in use at exit: 44,734 bytes in 92 blocks
==12582==   total heap usage: 5,503 allocs, 5,411 frees, 2,791,411 bytes allocated
==12582== 
==12582== 288 bytes in 1 blocks are possibly lost in loss record 18 of 27
==12582==    at 0x4C2FB55: calloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==12582==    by 0x40138B4: allocate_dtv (dl-tls.c:322)
==12582==    by 0x40138B4: _dl_allocate_tls (dl-tls.c:539)
==12582==    by 0x587026E: allocate_stack (allocatestack.c:588)
==12582==    by 0x587026E: pthread_create@@GLIBC_2.2.5 (pthread_create.c:539)
==12582==    by 0x479650: CreateThreadJoinable (pthreads_op.c:47)
==12582==    by 0x4796F1: CreateThread (pthreads_op.c:62)
==12582==    by 0x411CF9: w_create_output_threads (logcollector.c:1776)
==12582==    by 0x40DBB4: LogCollectorStart (logcollector.c:290)
==12582==    by 0x40B9A6: main (main.c:187)
==12582== 
==12582== 288 bytes in 1 blocks are possibly lost in loss record 19 of 27
==12582==    at 0x4C2FB55: calloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==12582==    by 0x40138B4: allocate_dtv (dl-tls.c:322)
==12582==    by 0x40138B4: _dl_allocate_tls (dl-tls.c:539)
==12582==    by 0x587026E: allocate_stack (allocatestack.c:588)
==12582==    by 0x587026E: pthread_create@@GLIBC_2.2.5 (pthread_create.c:539)
==12582==    by 0x479650: CreateThreadJoinable (pthreads_op.c:47)
==12582==    by 0x4796F1: CreateThread (pthreads_op.c:62)
==12582==    by 0x40DC23: LogCollectorStart (logcollector.c:301)
==12582==    by 0x40B9A6: main (main.c:187)
==12582== 
==12582== 1,152 bytes in 4 blocks are possibly lost in loss record 21 of 27
==12582==    at 0x4C2FB55: calloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==12582==    by 0x40138B4: allocate_dtv (dl-tls.c:322)
==12582==    by 0x40138B4: _dl_allocate_tls (dl-tls.c:539)
==12582==    by 0x587026E: allocate_stack (allocatestack.c:588)
==12582==    by 0x587026E: pthread_create@@GLIBC_2.2.5 (pthread_create.c:539)
==12582==    by 0x479650: CreateThreadJoinable (pthreads_op.c:47)
==12582==    by 0x4796F1: CreateThread (pthreads_op.c:62)
==12582==    by 0x4127CF: w_create_input_threads (logcollector.c:2061)
==12582==    by 0x40DBBE: LogCollectorStart (logcollector.c:293)
==12582==    by 0x40B9A6: main (main.c:187)
==12582== 
==12582== LEAK SUMMARY:
==12582==    definitely lost: 0 bytes in 0 blocks
==12582==    indirectly lost: 0 bytes in 0 blocks
==12582==      possibly lost: 1,728 bytes in 6 blocks
==12582==    still reachable: 43,006 bytes in 86 blocks
==12582==         suppressed: 0 bytes in 0 blocks
==12582== Reachable blocks (those to which a pointer was found) are not shown.
==12582== To see them, rerun with: --leak-check=full --show-leak-kinds=all
```

**Comment by TJOSERAFAEL** [comment]
### Dr memory report

```
~~1408~~ Error #6: LEAK 20 direct bytes 0x02b44238-0x02b4424c + 0 indirect bytes
~~1408~~ # 0 replace_malloc                     [d:\drmemory_package\common\alloc_replace.c:2576]
~~1408~~ # 1 libwinpthread-1.dll!mutex_impl_init [/build/mingw-w64-_1w3Xm/mingw-w64-4.0.4/mingw-w64-libraries/winpthread
s/src/mutex.c:70]
~~1408~~ # 2 libwinpthread-1.dll!pthread_mutex_lock [/build/mingw-w64-_1w3Xm/mingw-w64-4.0.4/mingw-w64-libraries/winpthr
eads/src/mutex.c:99]
~~1408~~ # 3 OSRegex_Execute_ex                 [os_regex/os_regex_execute.c:90]
~~1408~~ # 4 OSRegex_Execute                    [os_regex/os_regex_execute.c:30]
~~1408~~ # 5 OS_Regex                           [os_regex/os_regex.c:30]
~~1408~~ # 6 check_pattern_expand               [logcollector/logcollector.c:1360]
~~1408~~ # 7 LogCollectorStart                  [logcollector/logcollector.c:110]
~~1408~~ # 8 local_start                        [win32/win_agent.c:344]
~~1408~~ # 9 main                               [win32/win_agent.c:104]
~~1408~~
~~1408~~ ERRORS FOUND:
~~1408~~       1 unique,    87 total unaddressable access(es)
~~1408~~       2 unique,  2256 total uninitialized access(es)
~~1408~~       0 unique,     0 total invalid heap argument(s)
~~1408~~       0 unique,     0 total GDI usage error(s)
~~1408~~       0 unique,     0 total handle leak(s)
~~1408~~       0 unique,     0 total warning(s)
~~1408~~       3 unique,    47 total,   1438 byte(s) of leak(s)
~~1408~~       0 unique,     0 total,      0 byte(s) of possible leak(s)
~~1408~~ ERRORS IGNORED:
~~1408~~     134 potential error(s) (suspected false positives)
~~1408~~          (details: C:\Users\Administrator\AppData\Roaming\Dr. Memory\DrMemory-ossec-agent.exe.872.000\potential
_errors.txt)
~~1408~~       2 potential leak(s) (suspected false positives)
~~1408~~          (details: C:\Users\Administrator\AppData\Roaming\Dr. Memory\DrMemory-ossec-agent.exe.872.000\potential
_errors.txt)
~~1408~~     426 unique,  2090 total, 485642 byte(s) of still-reachable allocation(s)
~~1408~~          (re-run with "-show_reachable" for details)
```

**Pull request merged**